### ISSUE TITLE
google-protobuf won't build on Linux after adding MinGW platform

### ISSUE THREAD

**Issue by deivid-rodriguez** [rubygems/rubygems#7321]
### Discussed in https://github.com/rubygems/rubygems/discussions/7315

<div type='discussions-op-text'>

<sup>Originally posted by **AnyOldName3** December 20, 2023</sup>
I had a working Jekyll project that built fine under both real Linux and WSL2, but for reasons not worth getting into, needed to switch from WSL2 to Windows. I ran `bundle lock --add-platform x64-mingw-ucrt01`, the results of which can be seen as this commit: https://github.com/AnyOldName3/osg-website-replacement/commit/a72328e722c0ab38b0914794e767443a14210fb6. The line I believe to be most important changes `google-protobuf (3.25.1-x86_64-linux)` to `google-protobuf (3.25.1)`, and looking at the contents of the two packages, it appears the former has prebuilt `.so`s whereas the latter is source-only, so the shared library needs building.

As intended, this made the project work with Windows Ruby, however, it also produced the following errors on Linux:
```
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /mnt/c/vsg/osg-website-replacement/vendor/bundler/ruby/3.0.0/gems/google-protobuf-3.25.1/ext/google/protobuf_c
/usr/bin/ruby3.0 -I /usr/lib/ruby/vendor_ruby -r ./siteconf20231220-407-i7czk0.rb extconf.rb
creating Makefile

current directory: /mnt/c/vsg/osg-website-replacement/vendor/bundler/ruby/3.0.0/gems/google-protobuf-3.25.1/ext/google/protobuf_c
make DESTDIR\= clean

current directory: /mnt/c/vsg/osg-website-replacement/vendor/bundler/ruby/3.0.0/gems/google-protobuf-3.25.1/ext/google/protobuf_c
make DESTDIR\=
compiling convert.c
convert.c: In function ‘Msgval_IsEqual’:
convert.c:295:58: warning: passing argument 4 of ‘shared_Msgval_IsEqual’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  295 |                                             type_info.def.msgdef, &status);
      |                                             ~~~~~~~~~~~~~^~~~~~~
In file included from convert.c:21:
shared_convert.h:19:60: note: expected ‘upb_MessageDef *’ but argument is of type ‘const upb_MessageDef *’
   19 |                            upb_CType type, upb_MessageDef* msgdef,
      |                                            ~~~~~~~~~~~~~~~~^~~~~~
convert.c:299:5: error: format not a string literal and no format arguments [-Werror=format-security]
  299 |     rb_raise(rb_eRuntimeError, upb_Status_ErrorMessage(&status));
      |     ^~~~~~~~
convert.c: In function ‘Msgval_GetHash’:
convert.c:308:41: warning: passing argument 3 of ‘shared_Msgval_GetHash’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  308 |       val, type_info.type, type_info.def.msgdef, seed, &status);
      |                            ~~~~~~~~~~~~~^~~~~~~
In file included from convert.c:21:
shared_convert.h:23:48: note: expected ‘upb_MessageDef *’ but argument is of type ‘const upb_MessageDef *’
   23 |                                upb_MessageDef* msgdef, uint64_t seed,
      |                                ~~~~~~~~~~~~~~~~^~~~~~
convert.c:312:5: error: format not a string literal and no format arguments [-Werror=format-security]
  312 |     rb_raise(rb_eRuntimeError, upb_Status_ErrorMessage(&status));
      |     ^~~~~~~~
cc1: some warnings being treated as errors
make: *** [Makefile:245: convert.o] Error 1

make failed, exit code 2
```
This log is from using WSL2, but it also happens in CI: https://github.com/AnyOldName3/osg-website-replacement/actions/runs/7266847247/job/19829238652. I believe that's Arch running in Docker (on an Ubuntu host) whereas WSL2 is running Ubuntu, so dependency packages from the OS should be different - they're at least using different versions of Ruby.

`-Werror=format-security` is coming from `rbconfig.rb` on the WSL2 system, and I imagine that's where it's coming from on the probably-Arch VM, too. Conceivably, other compiler flags requested by that file could be responsible for the other errors. I don't know whether this is a case of a broken gem (i.e. that these flags should work), a packaging error for Ruby itself (i.e. that these flags shouldn't have ended up in `rbconfig.rb`, or some other kind of problem. I also don't know how I'd go about working around this, e.g. by somehow forcing `Gemfile.lock` to use the prebuilt Linux variant of the package on Linux.</div>

**Comment by deivid-rodriguez** [comment]
I can't reproduce this. Under which platform/OS did you run `bundle lock --add-platform x64-mingw-ucrt`?

**Comment by AnyOldName3** [comment]
That was done under WSL2, running Ubuntu 22.04.3. The Bash login welcome message includes the line
```
Welcome to Ubuntu 22.04.3 LTS (GNU/Linux 5.15.133.1-microsoft-standard-WSL2 x86_64)
```
`ruby -v` prints
```
ruby 3.0.2p107 (2021-07-07 revision 0db68f0233) [x86_64-linux-gnu]
```
and `bundle -v` prints
```
Bundler version 2.4.22
```

**Comment by deivid-rodriguez** [comment]
Thanks! I can reproduce now using Ruby 3.0, let me have a closer look.

**Comment by deivid-rodriguez** [comment]
#7324 should fix this!

**Comment by AnyOldName3** [comment]
My understanding of that PR (as someone with extremely limited Ruby experience) is that had the bug not existed, I'd have ended up with something resembling
```
    google-protobuf (3.25.1)
    google-protobuf (3.25.1-x86_64-linux)
```
in my `Gemfile.lock`, and this would have caused the prebuilt variant of the gem to be used on x86_64 Linux and the C source variant everywhere else.

Manually making that change (as seen in this commit) https://github.com/AnyOldName3/osg-website-replacement/commit/1468850b26636e46dab53192b8f5a9ba78ef16ec seemed to work locally - `bundle install` succeeded both with native Windows (MinGW) Ruby, and native Ubuntu Ruby running via WSL2.

However, when I pushed that commit, I had the same build error as before in CI (which as mentioned previously, is Arch running via Docker), meaning it's still using the C source variant of the package in preference to the precompiled one. The log is here: https://github.com/AnyOldName3/osg-website-replacement/actions/runs/7399341072/job/20130528820. The most obvious difference I can find is that the GitHub Action I'm using to build the site is hardcoded to use Ruby 2.7, whereas on Windows I'm using 3.1.3 and on WSL2 it's got 3.0.2 as that's what Ubuntu 22.04 provides. I don't know if that's supposed to make a difference, but it works when the C source variant is excluded from the lockfile, so it's not simply that Ruby 2.7 is totally incompatible.

**Comment by deivid-rodriguez** [comment]
You probably need to run `bundle lock --add-platform aarch64-linux` for your lockfile to also work on that platform?

**Comment by AnyOldName3** [comment]
I don't need it to work on that platform. The build's failing on x86_64-linux, where it's supposed to be using the prebuilt gem.

**Comment by deivid-rodriguez** [comment]
Oh, sorry, I misunderstood,

Can you confirm this is still not working with Bundler 2.5.4?

**Comment by AnyOldName3** [comment]
To confirm, that means changing this `Gemfile.lock` https://github.com/AnyOldName3/osg-website-replacement/blob/1468850b26636e46dab53192b8f5a9ba78ef16ec/Gemfile.lock to say
```
BUNDLED WITH
   2.5.4
```
at the bottom and bundler will fetch and use the correct version of itself, and it should work (meaning use the C source protobuf gem everywhere except x86_64-linux, but the prebuilt x86_64-linux one instead on x86_64-linux)?

**Comment by deivid-rodriguez** [comment]
I'd hope so, yes!

**Comment by AnyOldName3** [comment]
```
ERROR:  Error installing bundler:
	There are no versions of bundler (= 2.5.4) compatible with your Ruby & RubyGems
	bundler requires Ruby version >= 3.0.0. The current ruby version is 2.7.8.225.
```

Note that https://bundler.io/compatibility.html doesn't mention that any version requires Ruby 3.

**Comment by AnyOldName3** [comment]
Again, to clarify, when I run things locally, it's on post-3 Ruby, it's just the CI that's set up to use 2.7.

**Comment by deivid-rodriguez** [comment]
Good catch, we need to update https://github.com/rubygems/bundler-site with the information for Bundler 2.5.

I'm afraid you're stuck with Bundler 2.4 because of https://github.com/jeffreytse/jekyll-deploy-action/issues/53.

An idea would be to switch to https://github.com/actions/starter-workflows/blob/c31fe3d5d44d7cb4c912f4c3213f7b4610f13ea2/pages/jekyll.yml, which seems to be using `ruby/setup-ruby` and thus can use any ruby version you like.

**Comment by AnyOldName3** [comment]
I think I avoided that one because it mentioned GitHub Pages before it built the site, so I was concerned it might impose the restrictions on Jekyll versions and plugins that were the whole reason I was moving to an explicit action. Anyway, I *think* I might be able to get around this going forward, as when I installed another gem locally I needed as a plugin dependency, it changed my `Gemfile.lock` pretty significantly, including changing 
```
google-protobuf (3.25.1)
```
to
```
    google-protobuf (3.25.1-x86_64-linux)
    google-protobuf (3.25.1-x86_64-mingw-ucrt)
```
so I'm hoping that means it'll use prebuilt gems on both platforms, and none will even see the C source one.

That change isn't ready to push yet, so I'll see how it goes when it is.
